deepset_cloud_sdk/workflows/pipeline_client/pipeline_service.py

```diff
@@ -0,0 +1,239 @@
+        else:
+            logger.debug(f"{method_name} method already exists on {class_name} class")
+
+    # Add methods to both Pipeline classes
+    add_method_if_not_exists(HaystackPipeline, "publish_async", publish_to_deepset_async, "Pipeline")
+    add_method_if_not_exists(HaystackPipeline, "publish", publish_to_deepset, "Pipeline")
+    add_method_if_not_exists(HaystackAsyncPipeline, "publish_async", publish_to_deepset_async, "AsyncPipeline")
+    add_method_if_not_exists(HaystackAsyncPipeline, "publish", publish_to_deepset, "AsyncPipeline")
+
+
+class PipelineService:
+    """Handles the publishing of Haystack pipelines and indexes to deepset AI platform."""
+
+    def __init__(self, api: DeepsetCloudAPI) -> None:
+        """Initialize the pipeline service.
+
+        :param api: An initialized DeepsetCloudAPI instance
+        """
+        self._api = api
+        self._yaml = YAML()
+        self._yaml.preserve_quotes = True
+        self._yaml.indent(mapping=2, sequence=2)
+
+    @classmethod
+    async def factory(cls, config: CommonConfig) -> "PipelineService":
+        """Create a new instance of the pipeline service.
+
+        :param config: CommonConfig object.
+        """
+        async with DeepsetCloudAPI.factory(config) as api:
+            return cls(api)
+
+    async def publish_async(self, pipeline: PipelineProtocol, config: IndexConfig | PipelineConfig) -> None:
+        """Publish a pipeline or index to deepset AI platform.
+
+        :param pipeline: The pipeline or index to publish. Must be a Haystack Pipeline or AsyncPipeline
+        :param config: Configuration for publishing, either `IndexConfig` or `PipelineConfig`.
+            If publishing an index, the config argument is expected to be of type `IndexConfig`,
+            if publishing a pipeline, the config argument is expected to be of type `PipelineConfig`.
+
+        :raises TypeError: If the pipeline object isn't a Haystack Pipeline or AsyncPipeline
+        :raises ValueError: If no workspace is configured
+        :raises ImportError: If haystack-ai is not installed
+        """
+        logger.debug(f"Starting async publishing for {config.name}")
+
+        try:
+            from haystack import AsyncPipeline as HaystackAsyncPipeline
+            from haystack import Pipeline as HaystackPipeline
+        except ImportError as err:
+            raise ImportError("Can't import Pipeline or AsyncPipeline, because haystack-ai is not installed.") from err
+
+        if not isinstance(pipeline, (HaystackPipeline, HaystackAsyncPipeline)):
+            raise TypeError(
+                "Haystack Pipeline or AsyncPipeline object expected. "
+                "Make sure you have installed haystack-ai and use Pipeline or AsyncPipeline "
+                "to define your pipeline or index."
+            )
+
+        if not DEFAULT_WORKSPACE_NAME:
+            raise ValueError(
+                "We couldn't find the workspace to publish to in your environment. "
```

Comment:
should it be: ..in your deepset organization?

Reply:
nah, it's actually from their environment configuration when they login using the sdk.